### ISSUE TITLE
NullPointerException in storedAndRequestedTypesMatch(Neo4jEntityConverterImpl.java:124) when using embed mode

### ISSUE THREAD

**Issue by liudonghua123** [spring-projects/spring-data-neo4j#273]
When I use embed mode of GraphDatabaseService, it give me some NPE error!
The relevant code I wrote is as following.
```java

	private static final String DATABASE_PATH = "resources/database_movie_recommend_neo4j.db.sm";
	private static final String IS_EMBEDED = System.getenv("NEO4J_IS_EMBEDED") != null ? System
			.getenv("NEO4J_IS_EMBEDED") : "true";
	private static final String NEO4J_URL = System.getenv("NEO4J_URL") != null ? System
			.getenv("NEO4J_URL") : "http://localhost:7474/db/data/";

	@Bean
	public GraphDatabaseService graphDatabaseService() {
		if (Boolean.parseBoolean(IS_EMBEDED)) {
			log.info("Using embed mode");
			return new GraphDatabaseFactory()
					.newEmbeddedDatabase(DATABASE_PATH);
		} else {
			log.info("Using rest http mode");
			return new SpringRestGraphDatabase(NEO4J_URL);
		}
	}

	public Map<String, Object> recommendationGraph(int userId, int limit) {
		List<Integer> userRatedMovieIds = movieRepository
				.findUserRatedMovieIds(userId);
		Map<Double, Movie> recommendations = getRecommendations(userId, limit);

		List<NodeInfo> nodeInfos = movieRepository
				.findRecommendationRelevantNodes(userId, userRatedMovieIds);
		List<EdgeInfo> edgeInfos = movieRepository
				.findRecommendationRelevantEdges(userId, userRatedMovieIds);

		return Utils.toGraphJSONFormat(nodeInfos, edgeInfos, recommendations);
	}

	private Map<Double, Movie> getRecommendations(int userId, int size) {
		int candidateSize = 100;
		List<Movie> recommendMovies = movieRepository.findRecommendMovies(userId, candidateSize);
		// some other code ommited here
	}

	@Query("match (u:User)-[:RATED]->(um:Movie) "
			+ "where u.id={0} "
			+ "with collect(distinct um.id) as ums "
			+ "match (u)-[:RATED]->(:Movie)<-[:RATED]-(ou:User)-[:RATED]->(om:Movie) "
			+ "where not om.id in ums "
			+ "return distinct om "
			+ "limit {1}")
	List<Movie> findRecommendMovies(int userid, int limit);
```

However I do not get such exception when I use rest http mode.

The following is stack track of exception
```
java.lang.NullPointerException: null
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityConverterImpl.storedAndRequestedTypesMatch(Neo4jEntityConverterImpl.java:124)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityConverterImpl.read(Neo4jEntityConverterImpl.java:88)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityPersister$CachedConverter.read(Neo4jEntityPersister.java:170)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityPersister.createEntityFromState(Neo4jEntityPersister.java:192)
	at org.springframework.data.neo4j.support.Neo4jTemplate.createEntityFromState(Neo4jTemplate.java:224)
	at org.springframework.data.neo4j.fieldaccess.GraphBackedEntityIterableWrapper.underlyingObjectToObject(GraphBackedEntityIterableWrapper.java:41)
	at org.springframework.data.neo4j.fieldaccess.GraphBackedEntityIterableWrapper.underlyingObjectToObject(GraphBackedEntityIterableWrapper.java:27)
	at org.neo4j.helpers.collection.IterableWrapper$MyIteratorWrapper.underlyingObjectToObject(IterableWrapper.java:57)
	at org.neo4j.helpers.collection.IteratorWrapper.next(IteratorWrapper.java:47)
	at org.neo4j.helpers.collection.IteratorUtil.addToCollection(IteratorUtil.java:388)
	at org.neo4j.helpers.collection.IteratorUtil.addToCollection(IteratorUtil.java:467)
	at org.springframework.data.neo4j.fieldaccess.RelatedToViaCollectionFieldAccessorFactory$RelatedToViaCollectionFieldAccessor.getValue(RelatedToViaCollectionFieldAccessorFactory.java:122)
	at org.springframework.data.neo4j.fieldaccess.DefaultEntityState.getValue(DefaultEntityState.java:97)
	at org.springframework.data.neo4j.support.mapping.SourceStateTransmitter.copyEntityStatePropertyValue(SourceStateTransmitter.java:87)
	at org.springframework.data.neo4j.support.mapping.SourceStateTransmitter.access$000(SourceStateTransmitter.java:37)
	at org.springframework.data.neo4j.support.mapping.SourceStateTransmitter$2.doWithAssociation(SourceStateTransmitter.java:58)
	at org.springframework.data.mapping.model.BasicPersistentEntity.doWithAssociations(BasicPersistentEntity.java:339)
	at org.springframework.data.neo4j.support.mapping.SourceStateTransmitter.copyPropertiesFrom(SourceStateTransmitter.java:54)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityConverterImpl.loadEntity(Neo4jEntityConverterImpl.java:114)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityConverterImpl.read(Neo4jEntityConverterImpl.java:106)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityPersister$CachedConverter.read(Neo4jEntityPersister.java:170)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityPersister.createEntityFromState(Neo4jEntityPersister.java:189)
	at org.springframework.data.neo4j.support.mapping.Neo4jEntityPersister.projectTo(Neo4jEntityPersister.java:216)
	at org.springframework.data.neo4j.support.Neo4jTemplate.projectTo(Neo4jTemplate.java:241)
	at org.springframework.data.neo4j.support.conversion.EntityResultConverter.doConvert(EntityResultConverter.java:73)
	at org.springframework.data.neo4j.conversion.DefaultConverter.convert(DefaultConverter.java:44)
	at org.springframework.data.neo4j.support.conversion.EntityResultConverter.convert(EntityResultConverter.java:165)
	at org.springframework.data.neo4j.conversion.QueryResultBuilder$1.underlyingObjectToObject(QueryResultBuilder.java:86)
	at org.neo4j.helpers.collection.IterableWrapper$MyIteratorWrapper.underlyingObjectToObject(IterableWrapper.java:57)
	at org.neo4j.helpers.collection.IteratorWrapper.next(IteratorWrapper.java:47)
	at org.neo4j.helpers.collection.IteratorUtil.addToCollection(IteratorUtil.java:388)
	at org.neo4j.helpers.collection.IteratorUtil.addToCollection(IteratorUtil.java:467)
	at org.springframework.data.neo4j.conversion.ContainerConverter.toContainer(ContainerConverter.java:42)
	at org.springframework.data.neo4j.conversion.QueryResultBuilder.as(QueryResultBuilder.java:96)
	at org.springframework.data.neo4j.repository.query.GraphRepositoryQuery.dispatchQuery(GraphRepositoryQuery.java:118)
	at org.springframework.data.neo4j.repository.query.GraphRepositoryQuery$1.doWithGraph(GraphRepositoryQuery.java:90)
	at org.springframework.data.neo4j.support.Neo4jTemplate.doExecute(Neo4jTemplate.java:465)
	at org.springframework.data.neo4j.support.Neo4jTemplate.access$000(Neo4jTemplate.java:87)
	at org.springframework.data.neo4j.support.Neo4jTemplate$2.doInTransaction(Neo4jTemplate.java:479)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at org.springframework.data.neo4j.support.Neo4jTemplate.exec(Neo4jTemplate.java:476)
	at org.springframework.data.neo4j.repository.query.GraphRepositoryQuery.execute(GraphRepositoryQuery.java:84)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:393)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:371)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:136)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy108.findRecommendMovies(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy112.findRecommendMovies(Unknown Source)
	at com.liudonghua.apps.movie_recommendation_demo.service.MovieService.getRecommendations(MovieService.java:86)
	at com.liudonghua.apps.movie_recommendation_demo.service.MovieService.recommendationGraph(MovieService.java:76)
	at com.liudonghua.apps.movie_recommendation_demo.service.MovieService$$FastClassBySpringCGLIB$$61d601e5.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:717)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653)
	at com.liudonghua.apps.movie_recommendation_demo.service.MovieService$$EnhancerBySpringCGLIB$$8cc58b1.recommendationGraph(<generated>)
	at com.liudonghua.apps.movie_recommendation_demo.MovieApplication.recommendationGraph(MovieApplication.java:64)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:110)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:776)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:705)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:966)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:857)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:842)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:85)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:516)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1086)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:659)
	at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:223)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1558)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1515)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Unknown Source)
```

ps: The version of my neo4j database is 2.2.0. And in order to use neo4j 2.2.0 with spring-data-neo4j, I have to wrote some hack dependence in my pom.xml
```xml
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>1.2.3.RELEASE</version>
    </parent>

    <dependencies>
		<dependency>
			<groupId>org.neo4j</groupId>
			<artifactId>neo4j</artifactId>
			<version>2.2.0</version>
		</dependency>
		<dependency>
			<groupId>org.springframework.data</groupId>
			<artifactId>spring-data-commons</artifactId>
			<version>1.10.0.RELEASE</version>
		</dependency>
		<!-- Fix ClassNotFoundException: org.springframework.data.neo4j.core.UpdateableState -->
		<dependency>
			<groupId>org.springframework.data</groupId>
			<artifactId>spring-data-neo4j</artifactId>
			<version>3.3.0.RELEASE</version>
		</dependency>
    
		<dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-rest</artifactId>
        </dependency>
		
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-neo4j-rest</artifactId>
            <version>3.3.0.RELEASE</version>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-rest</artifactId>
        </dependency>
        
        <dependency>
            <groupId>org.hibernate</groupId>
            <artifactId>hibernate-validator</artifactId>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>
```

Any ideas how to solve this issue?

**Comment by KimJejun** [comment]
is it fixed? 
I met same exception

**Comment by KimJejun** [comment]
In java config.
If using NoopRelationshipTypeRepresentationStrategy to relationshipTypeRepresentationStrategy it makes this error in readAliasFrom(this method return null)

if using default relationshipTypeRepresentationStrategy(IndexBasedRelationshipTypeRepresentationStrategy) it can not save relation properties. save this (__type__ : Role)

if using NoopRelationshipTypeRepresentationStrategy without Transactional. it work with exceptions but relation properties saved.(ex, role : main actor)

It will be update? or make new relationTypeRepresentationStrategy?

**Comment by agupta1989** [comment]
Is this issue resolved ? I am using SDN 3.4.1.RELEASE with Spring Boot 1.3.0.RELEASE and Spring Cloud Brixton.M3. Whenever I save the relationship it throws me NPE. Although relationship get created. any work around ?

**Comment by mangrish** [comment]
Closed. Issue out of date.
